hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieHFileKeyExcludedDataBlock.java

```diff
@@ -0,0 +1,125 @@
+ * under the License.
+ */
+
+package org.apache.hudi.common.table.log.block;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.io.hfile.CacheConfig;
+import org.apache.hudi.common.model.HoodieLogFile;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ValidationUtils;
+import org.apache.hudi.io.storage.HoodieHFileKeyExcludedReader;
+import org.apache.hudi.io.storage.HoodieHFileReader;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * HFile data block for the Metadata table records. Since the backing log format for metadata table
+ * is the HFile KeyValue and since the key field in the metadata record payload is a duplicate
+ * of the Key in the Cell, the redundant key field in the record can be nullified to save on the
+ * cost. Such trimmed metadata records need to re-materialized with the key field during deserialization.
+ */
+public class HoodieHFileKeyExcludedDataBlock extends HoodieHFileDataBlock {
+
+  private static final Logger LOG = LogManager.getLogger(HoodieHFileKeyExcludedDataBlock.class);
+
+  public HoodieHFileKeyExcludedDataBlock(HoodieLogFile logFile, FSDataInputStream inputStream, Option<byte[]> content,
+                                         boolean readBlockLazily, long position, long blockSize, long blockEndPos,
+                                         Schema readerSchema, Map<HeaderMetadataType, String> header,
+                                         Map<HeaderMetadataType, String> footer, boolean enableInlineReading,
+                                         String keyField) {
+    super(logFile, inputStream, content, readBlockLazily, position, blockSize, blockEndPos, readerSchema, header,
+        footer, enableInlineReading, keyField);
+  }
+
+  public HoodieHFileKeyExcludedDataBlock(@NotNull List<IndexedRecord> records,
+                                         @NotNull Map<HeaderMetadataType, String> header, String keyField) {
+    super(records, header, keyField);
+  }
+
+  /**
+   * Serialize the metadata table record to byte buffer after any field trimming if needed.
+   *
+   * @param record         - Record to serialize
+   * @param schemaKeyField - Key field in the schema
+   * @return Serialized byte array of the metadata trimmed record
+   */
+  @Override
+  protected ByteBuffer serializeRecord(final IndexedRecord record, final Option<Schema.Field> schemaKeyField) {
+    if (!schemaKeyField.isPresent()) {
+      return super.serializeRecord(record, schemaKeyField);
+    }
+
+    ValidationUtils.checkArgument(record.getSchema() != null, "Unknown schema for the record!");
```

Comment:
do we need to validate for every record? any cost associated with this

Reply:
removed it.